client.go

```diff
@@ -448,8 +446,8 @@ func (c *Client) connect(identifier, serverAddr string) error {
 	c.log.Debug("client has started successfully.")
 	c.redialBackoff.Reset() // we successfully connected, so we can reset the backoff
 
-	c.mu.Lock()
-	if c.startNotify != nil && !c.closed {
+	c.closedMu.RLock()
+	if !c.closed && c.startNotify != nil {
```

Comment:
you could use c.isClosed() here since we are not touching to "closed" within if statement anymore